README.md
# InnoSetupScripts
ISS using Various Pascal(Inno Setup) Scripts

## Donation

Soon


Src Code/history.txt
Date        Version  New features compared to previous version

2007-04-16  0.1      Bytecoder, bitcoder and 2^n semi-dynamic RangeCoder
                     Greedy parsing
                     Only 4+ byte matches
                     Hashtable matchfinders: simple and caching
                     Blockwise compression

2007-05-11  0.2      Huffman coder
                     Lazy parsing
                     3-byte matches
                     Dropping too far matches
                     Sliding window
                     Gzip-like cmdline interface

2007-07-25  0.25     Repdist, repboth and repchar codes
                     2-byte matches
                     Improved lz parsing
                     Delta filter

2008-03-13  0.3      Huffman decoder with variable block size
                     Tested on 100 gb of data and fixed all bugs found

2008-04-08  0.4      Cycling matchfinder
                     Large, updated at each step hashes for 2/3-byte strings
                     More interactive user interface

2008-06-02  0.4a     Fixed bug in Huffman coder: unstable sorting
                     Win32 & Linux versions

2008-10-28  0.4b     Added -11:c1 .. -11:c3 modes

2014-01-15  0.5      Removed modes -6 -8 -10 -12, renumerated -7/9/11 to -6/7/8
                     -7/-8: 1% better compression and 5-20% better speed due to use of ht5
                     New -9..-11 modes: 2gb hash and ht5..ht7 match finders
                     In small tornado.exe -c1..-c3 available for -7..-11 modes
                     Support for -l not power of 2   (in all matchfinders?)
                     Options -ah, -al, -t#, -cpu, -q, -delete, stats->stderr
                     Checks at decoding in order to prevent segfaults
                     MSVC/ICL compatibility, use of wall-clock times
                     Linux: 64-bit file offsets!
                     PowerPC (big-endian cpus) compatibility, thanks to Joachim Henke
                     64-bit compilers support!
                     -3 mode compression improved by 2-3%
                     2mb hash in -4/-5 modes improved speed by 20-30%
                     Fixed long-standing decompression bug (table split across 2 write chunks)
                     Removed CHECK_FOR_DATA_TABLE(2) in -3/-4 modes


2014-03-08  0.6      http://freearc.org/download/research/tornado06.zip

Compression:
* new -11..-16 modes employing the new optimal parser (-p4)
* -fb (FAST BYTES), new parameter used only by the optimal parser
* old -10 mode was removed, old -11 mode was better tuned and renamed to -10
* -s1 option meaning changed to "64kb hash3 + 4kb hash2", slightly improving compression in -5/-6 modes
* new chash5..7 (-x15..17) and bt4..7 (-x24..27) match finders
* all chash/cchash/bt matchfinders support -l values that is not power of 2
* new manual.txt explaining all compression parameters

Command line and progress indicator:
* -slp/-rem options, large pages are allocated by default
* Win7 taskbar progress indication (green bar) plus info in the console window title
* returns Errcode 2 on any error, clears the window title on ^Break
* prints ranges and default values for each option, checks option correctness
* on Windows, prohibits simultaneous writing by several compression processes to the same output file
* prints byte-exact filesizes; speeds are measured in MiB/s
* progress indicator is more accurate and updated only once per 0.2 seconds

Compilation:
* on Windows, compile.cmd supports many GCC/MSVC/ICL versions
* on Linux, compile.sh now produces executables that can process files larger than 2 GB


2014-03-21  0.6a     http://freearc.org/download/research/tornado06a.zip

* FreeBSD support developed by m^2
* fixed bugs in -11..16 and -10 -x24 modes
* fixed bug in the decompression progress indicator code
* prohibited use of -t1 with incompatible -p4 and -x2* modes


Src Code/c/3dtest/3dtest.c
/*
	BASS 3D test
	Copyright (c) 1999-2017 Un4seen Developments Ltd.
*/

#include <windows.h>
#include <commctrl.h>
#include <math.h>
#include <stdio.h>
#include "bass.h"

HWND win = NULL;

// channel (sample/music) info structure
typedef struct {
	DWORD channel;			// the channel
	BASS_3DVECTOR pos, vel;	// position,velocity
} Channel;

Channel *chans = NULL;		// the channels
int chanc = 0, chan = -1;	// number of channels, current channel

#define TIMERPERIOD	50		// timer period (ms)
#define MAXDIST		50		// maximum distance of the channels (m)
#define SPEED		12		// speed of the channels' movement (m/s)

// Display error dialogs
void Error(const char *es)
{
	char mes[200];
	sprintf(mes, "%s\n(error code: %d)", es, BASS_ErrorGetCode());
	MessageBox(win, mes, 0, 0);
}

// Messaging macros
#define ITEM(id) GetDlgItem(win,id)
#define MESS(id,m,w,l) SendDlgItemMessage(win,id,m,(WPARAM)(w),(LPARAM)(l))
#define LM(m,w,l) MESS(10,m,w,l)
#define EM(m,w,l) MESS(22,m,w,l)

void UpdateDisplay()
{
	HDC dc;
	RECT r;
	int c, x, y, cx, cy;
	HBRUSH red = CreateSolidBrush(0xff);

	HWND w = ITEM(30);
	dc = GetDC(w);
	GetClientRect(w, &r);
	cx = r.right / 2;
	cy = r.bottom / 2;

	// clear the display
	FillRect(dc, &r, (HBRUSH)GetStockObject(WHITE_BRUSH));

	// Draw the listener
	SelectObject(dc, GetStockObject(GRAY_BRUSH));
	Ellipse(dc, cx - 4, cy - 4, cx + 4, cy + 4);

	for (c = 0; c < chanc; c++) {
		// If the channel's playing then update it's position
		if (BASS_ChannelIsActive(chans[c].channel) == BASS_ACTIVE_PLAYING) {
			// Check if channel has reached the max distance
			if (chans[c].pos.z >= MAXDIST || chans[c].pos.z <= -MAXDIST)
				chans[c].vel.z = -chans[c].vel.z;
			if (chans[c].pos.x >= MAXDIST || chans[c].pos.x <= -MAXDIST)
				chans[c].vel.x = -chans[c].vel.x;
			// Update channel position
			chans[c].pos.z += chans[c].vel.z * TIMERPERIOD / 1000;
			chans[c].pos.x += chans[c].vel.x * TIMERPERIOD / 1000;
			BASS_ChannelSet3DPosition(chans[c].channel, &chans[c].pos, NULL, &chans[c].vel);
		}
		// Draw the channel position indicator
		x = cx + (int)((cx - 10) * chans[c].pos.x / MAXDIST);
		y = cy - (int)((cy - 10) * chans[c].pos.z / MAXDIST);
		SelectObject(dc, chan == c ? red : GetStockObject(WHITE_BRUSH));
		Ellipse(dc, x - 4, y - 4, x + 4, y + 4);
	}
	// Apply the 3D changes
	BASS_Apply3D();

	ReleaseDC(w, dc);
	DeleteObject(red);
}

// Update the button states
void UpdateButtons()
{
	int a;
	for (a = 12; a <= 17; a++)
		EnableWindow(ITEM(a), chan == -1 ? FALSE : TRUE);
	if (chan != -1) {
		SetDlgItemInt(win, 15, abs((int)chans[chan].vel.x), FALSE);
		SetDlgItemInt(win, 16, abs((int)chans[chan].vel.z), FALSE);
	}
}

INT_PTR CALLBACK dialogproc(HWND h, UINT m, WPARAM w, LPARAM l)
{
	static OPENFILENAME ofn;

	switch (m) {
		case WM_TIMER:
			UpdateDisplay();
			break;

		case WM_COMMAND:
			switch (LOWORD(w)) {
				case 10: // change the selected channel
					if (HIWORD(w) != LBN_SELCHANGE) break;
					chan = LM(LB_GETCURSEL, 0, 0);
					if (chan == LB_ERR) chan = -1;
					UpdateButtons();
					break;
				case 11: // add a channel
					{
						char file[MAX_PATH] = "";
						DWORD newchan;
						ofn.lpstrFile = file;
						if (GetOpenFileName(&ofn)) {
							// Load a music or sample from "file"
							if ((newchan = BASS_MusicLoad(FALSE, file, 0, 0, BASS_MUSIC_RAMP | BASS_SAMPLE_LOOP | BASS_SAMPLE_3D, 1))
								|| (newchan = BASS_SampleLoad(FALSE, file, 0, 0, 1, BASS_SAMPLE_LOOP | BASS_SAMPLE_3D | BASS_SAMPLE_MONO))) {
								Channel *c;
								chanc++;
								chans = (Channel*)realloc((void*)chans, chanc * sizeof(Channel));
								c = chans + chanc - 1;
								memset(c, 0, sizeof(Channel));
								c->channel = newchan;
								BASS_SampleGetChannel(newchan, FALSE); // initialize sample channel
								LM(LB_ADDSTRING, 0, strrchr(file, '\\') + 1);
							} else
								Error("Can't load file (note samples must be mono)");
						}
					}
					break;
				case 12: // remove a channel
					{
						Channel *c = chans + chan;
						BASS_SampleFree(c->channel);
						BASS_MusicFree(c->channel);
						memmove(c, c + 1, (chanc - chan - 1) * sizeof(Channel));
						chanc--;
						LM(LB_DELETESTRING, chan, 0);
						chan = -1;
						UpdateButtons();
					}
					break;
				case 13:
					BASS_ChannelPlay(chans[chan].channel, FALSE);
					break;
				case 14:
					BASS_ChannelPause(chans[chan].channel);
					break;
				case 15: // X velocity
					if (HIWORD(w) == EN_CHANGE) {
						int v = GetDlgItemInt(win, 15, 0, FALSE);
						if (abs((int)chans[chan].vel.x) != v) chans[chan].vel.x = v;
					}
					break;
				case 16: // Z velocity
					if (HIWORD(w) == EN_CHANGE) {
						int v = GetDlgItemInt(win, 16, 0, FALSE);
						if (abs((int)chans[chan].vel.z) != v) chans[chan].vel.z = v;
					}
					break;
				case 17: // reset the position and velocity to 0
					memset(&chans[chan].pos, 0, sizeof(chans[chan].pos));
					memset(&chans[chan].vel, 0, sizeof(chans[chan].vel));
					UpdateButtons();
					break;
				case IDCANCEL:
					DestroyWindow(h);
					break;
			}
			break;

		case WM_HSCROLL:
			if (l) {
				int pos = SendMessage((HWND)l, TBM_GETPOS, 0, 0);
				switch (GetDlgCtrlID((HWND)l)) {
					case 20: // change the rolloff factor
						BASS_Set3DFactors(-1, pow(2, (pos - 10) / 5.0), -1);
						break;
					case 21: // change the doppler factor
						BASS_Set3DFactors(-1, -1, pow(2, (pos - 10) / 5.0));
						break;
				}
			}
			break;

		case WM_INITDIALOG:
			win = h;

			EM(CB_ADDSTRING, 0, "Off");
			EM(CB_ADDSTRING, 0, "Generic");
			EM(CB_ADDSTRING, 0, "Padded Cell");
			EM(CB_ADDSTRING, 0, "Room");
			EM(CB_ADDSTRING, 0, "Bathroom");
			EM(CB_ADDSTRING, 0, "Living Room");
			EM(CB_ADDSTRING, 0, "Stone Room");
			EM(CB_ADDSTRING, 0, "Auditorium");
			EM(CB_ADDSTRING, 0, "Concert Hall");
			EM(CB_ADDSTRING, 0, "Cave");
			EM(CB_ADDSTRING, 0, "Arena");
			EM(CB_ADDSTRING, 0, "Hangar");
			EM(CB_ADDSTRING, 0, "Carpeted Hallway");
			EM(CB_ADDSTRING, 0, "Hallway");
			EM(CB_ADDSTRING, 0, "Stone Corridor");
			EM(CB_ADDSTRING, 0, "Alley");
			EM(CB_ADDSTRING, 0, "Forest");
			EM(CB_ADDSTRING, 0, "City");
			EM(CB_ADDSTRING, 0, "Mountains");
			EM(CB_ADDSTRING, 0, "Quarry");
			EM(CB_ADDSTRING, 0, "Plain");
			EM(CB_ADDSTRING, 0, "Parking Lot");
			EM(CB_ADDSTRING, 0, "Sewer Pipe");
			EM(CB_ADDSTRING, 0, "Under Water");
			EM(CB_ADDSTRING, 0, "Drugged");
			EM(CB_ADDSTRING, 0, "Dizzy");
			EM(CB_ADDSTRING, 0, "Psychotic");
			EM(CB_SETCURSEL, 0, 0);

			MESS(20, TBM_SETRANGE, FALSE, MAKELONG(0, 20));
			MESS(20, TBM_SETPOS, TRUE, 10);
			MESS(21, TBM_SETRANGE, FALSE, MAKELONG(0, 20));
			MESS(21, TBM_SETPOS, TRUE, 10);

			SetTimer(h, 1, TIMERPERIOD, NULL);
			memset(&ofn, 0, sizeof(ofn));
			ofn.lStructSize = sizeof(ofn);
			ofn.hwndOwner = h;
			ofn.nMaxFile = MAX_PATH;
			ofn.Flags = OFN_HIDEREADONLY | OFN_EXPLORER;
			ofn.lpstrFilter = "wav/aif/mo3/xm/mod/s3m/it/mtm/umx\0*.wav;*.aif;*.mo3;*.xm;*.mod;*.s3m;*.it;*.mtm;*.umx\0"
				"All files\0*.*\0\0";
			return 1;

		case WM_DESTROY:
			KillTimer(h, 1);
			if (chans) free(chans);
			PostQuitMessage(0);
			break;
	}
	return 0;
}

int PASCAL WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow)
{
	MSG msg;

	// check the correct BASS was loaded
	if (HIWORD(BASS_GetVersion()) != BASSVERSION) {
		MessageBox(0, "An incorrect version of BASS.DLL was loaded", 0, MB_ICONERROR);
		return 0;
	}

	{ // enable trackbar support
		INITCOMMONCONTROLSEX cc = { sizeof(cc),ICC_BAR_CLASSES };
		InitCommonControlsEx(&cc);
	}

	// Create the main window
	if (!CreateDialog(hInstance, MAKEINTRESOURCE(1000), NULL, dialogproc)) {
		Error("Can't create window");
		return 0;
	}

	// Initialize the default output device with 3D support
	if (!BASS_Init(-1, 44100, BASS_DEVICE_3D, win, NULL)) {
		Error("Can't initialize output device");
		DestroyWindow(win);
		return 0;
	}

	// Use meters as distance unit, real world rolloff, real doppler effect
	BASS_Set3DFactors(1, 1, 1);

	while (GetMessage(&msg, NULL, 0, 0) > 0) {
		if (!IsDialogMessage(win, &msg)) {
			TranslateMessage(&msg);
			DispatchMessage(&msg);
		}
	}

	BASS_Free();

	return 0;
}
